README.md
[![Build Status](https://travis-ci.org/stjhimy/rack-www.svg?branch=master)](https://travis-ci.org/stjhimy/rack-www)

# rack-www

Rack middleware to force subdomain redirects, e.g: www or www2

## Installation

```ruby
  gem 'rack-www'
```

## Usage

Usage (by default it will redirect all requests to www subdomain):

```ruby
  config.middleware.use Rack::WWW
```

Customizing the :www option to true or false:

```ruby
  #redirects all traffic to www
  config.middleware.use Rack::WWW, www: true


  #redirects all traffic to the same domain without www
  config.middleware.use Rack::WWW, www: false
```

Redirecting to a given subdomain:

```ruby
  config.middleware.use Rack::WWW, :subdomain => "secure"
```

If you like it's also possible to show a message while redirecting the user:

```ruby
  config.middleware.use Rack::WWW, :www => false, :message => "You are being redirected..."
```

You can optionally specify predicate to determine if redirect should take place:

```ruby
  config.middleware.use Rack::WWW, :predicate => lambda { |env|
    !Rack::Request.new(env).params.has_key? "noredirect"
  }
```

Specifying host-regex to determine on what hosts to redirect:

```ruby
  config.middleware.use Rack::WWW, :host_regex => /example/i
```

This will only redirect when the host matches `example`, such as `example.com` or `example1.com`. It won't redirect on `localhost` or any other host that does not match the regex

It ignores any redirects when using IP addresses.

## Options

- :www => default is true, redirects all traffic to www;
- :subdomain => redirects to any given subdomain;
- :message => display a message while redirecting;
- :host_regex => redirects if the given host match

## License

MIT License. Copyright Jhimy Fernandes Villar http://stjhimy.com


CHANGELOG.md
Changelog
=========

Version 2.3.0 (July 17, 2017)
-----------------------------

* Allow host-regex option

Version 2.2.0 (August 11, 2016)
-----------------------------

* Do not append port for SSL redirect

Version 2.1.0
-----------------------------

* Ignore rack-www when request host is an IP address

Version 2.0.0
-----------------------------

* Update EVERYTHING

Version 1.5
-----------------------------

* Bug fixes

Version 1.4
-----------------------------

* Respect server port
* Body responds to :each
* Rack::Lint on tests

Version 1.3
-----------------------------

* Added possibility to redirect to any given :subdomain [Ryan Weald (https://github.com/rweald)]
* Added more tests

Version 1.2
-----------------------------

* Redirects to the right url without calling the app
* Keep the path when redirecting
* Keep the query string when redirecting
* Added more tests

Version 1.1
-----------------------------

* Added possibility to redirects with www or without www
* Added possibility to set a param :message to show while redirecting

Version 1.0
-----------------------------

* Redirects all traffic to www


test/helper.rb
require 'rack/www'
require 'minitest/autorun'
require 'rack/test'

class TestClass < MiniTest::Test
  include Rack::Test::Methods

  def default_app
    lambda do |_env|
      headers = { 'Content-Type' => 'text/html' }
      [200, headers, ['default body']]
    end
  end

  def app
    @app ||= Rack::Lint.new(Rack::WWW.new(default_app))
  end

  attr_writer :app
end


test/test_www.rb
require "rubygems"
require "rack/www"
require 'test/unit'
require 'active_support'
require 'rack/test'

class TestWWW < ActiveSupport::TestCase
  include Rack::Test::Methods

  def default_app
    lambda { |env|
      headers = {'Content-Type' => "text/html"}
      headers['Set-Cookie'] = "id=1; path=/\ntoken=abc; path=/; secure; HttpOnly"
      [200, headers, ["default body"]]
    }
  end

  def app
    @app ||= Rack::WWW.new(default_app)
  end
  attr_writer :app

  test "response has status 200[ok] in the default request" do
    get "http://www.example.com"
    assert_equal last_response.status, 200
  end

  test "response has status 301[redirects] when not a www url" do
    get "http://example.com"
    assert_equal last_response.status, 301
  end

  test "Redirects to right location when it's not www url" do
    get "http://example.com/"
    assert_equal "http://www.example.com/", last_response.headers['Location']
  end

  test "Redirects to a www url when param :www => true" do
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/"
    assert_equal "http://www.example.com/", last_response.headers['Location']
  end

  test "Redirects to a www url and keep the right path when param :www => true" do
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/path/1"
    assert_equal "http://www.example.com/path/1", last_response.headers['Location']
  end

  test "Redirects to a www url and keep the right query string when param :www => true" do
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/path/1?param=test"
    assert_equal "http://www.example.com/path/1?param=test", last_response.headers['Location']
  end

  test "Redirects to a non www url when param :www => false" do
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://www.example.com/"
    assert_equal "http://example.com/", last_response.headers['Location']
  end

  test "Redirects to a non  www url and keep the right path when param :www => false" do
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://www.example.com/path/1"
    assert_equal "http://example.com/path/1", last_response.headers['Location']
  end

  test "Redirects to a non  www url and keep the right query string when param :www => false" do
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://www.example.com/path/1?param=test"
    assert_equal "http://example.com/path/1?param=test", last_response.headers['Location']
  end
  
  test "Keeps the same url when non www url and param :www => false" do
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://example.com/"
    assert last_response.ok?
  end

  test "Changes the body content when param :message" do
    self.app = Rack::WWW.new(default_app, :www => true, :message => "redirecting now!")
    get "http://example.com/"
    assert_equal last_response.body, "redirecting now!"
  end

  test "Keeps the body empty when there's not a :message param" do
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/"
    assert_equal last_response.body, ""
  end

  test 'allow custom subdomain' do
    self.app = Rack::WWW.new(default_app, :subdomain => "secure")
    get 'http://example.com'
    assert_equal 'http://secure.example.com/', last_response.headers['Location']
  end

  test "custom subdomain with path" do
    self.app = Rack::WWW.new(default_app, :subdomain => "secure")
    get "http://example.com/path/1"
    assert_equal "http://secure.example.com/path/1", last_response.headers['Location']
  end

  test "custom subdomain with path and www" do
    self.app = Rack::WWW.new(default_app, :subdomain => "secure")
    get "http://www.example.com/path/1"
    assert_equal "http://secure.example.com/path/1", last_response.headers['Location']
  end

  test "custom subdomain with query string" do
    self.app = Rack::WWW.new(default_app, :subdomain => "secure")
    get "http://example.com/path/1?test=true"
    assert_equal "http://secure.example.com/path/1?test=true", last_response.headers['Location']
  end

  test "custom subdomain with query string and www" do
    self.app = Rack::WWW.new(default_app, :subdomain => "secure")
    get "http://www.example.com/path/1?test=true"
    assert_equal "http://secure.example.com/path/1?test=true", last_response.headers['Location']
  end

end


test/www_test.rb
require "helper"

class WWW < TestWWW
  def test_response_200
    get "http://www.example.com"
    assert_equal last_response.status, 200
  end

  def test_response_301
    get "http://example.com"
    assert_equal last_response.status, 301
  end

  def test_redirects
    get "http://example.com/"
    assert_equal "http://www.example.com/", last_response.headers['Location']
  end

  def test_www_true
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/"
    assert_equal "http://www.example.com/", last_response.headers['Location']
  end

  def test_www_true_with_path
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/path/1"
    assert_equal "http://www.example.com/path/1", last_response.headers['Location']
  end

  def test_www_true_with_query
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/path/1?param=test"
    assert_equal "http://www.example.com/path/1?param=test", last_response.headers['Location']
  end

  def test_www_false
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://www.example.com/"
    assert_equal "http://example.com/", last_response.headers['Location']
  end

  def test_www_false_with_path
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://www.example.com/path/1"
    assert_equal "http://example.com/path/1", last_response.headers['Location']
  end

  def test_www_false_with_query
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://www.example.com/path/1?param=test"
    assert_equal "http://example.com/path/1?param=test", last_response.headers['Location']
  end

  def test_www_false_non_www
    self.app = Rack::WWW.new(default_app, :www => false)
    get "http://example.com/"
    assert last_response.ok?
  end

  def test_body_with_message
    self.app = Rack::WWW.new(default_app, :www => true, :message => "redirecting now!")
    get "http://example.com/"
    assert_equal last_response.body, "redirecting now!"
  end

  def test_body_without_message
    self.app = Rack::WWW.new(default_app, :www => true)
    get "http://example.com/"
    assert_equal last_response.body, ""
  end

  def test_server_port
    get "http://example.com:8080/"
    assert_equal "http://www.example.com:8080/",
    last_response.headers['Location']
  end

  def test_server_port_80
    get "http://example.com:80/"
    assert_equal "http://www.example.com/",
    last_response.headers['Location']
  end

  def test_predicate_false
    self.app = Rack::WWW.new(default_app, :predicate => lambda { |env| false } )
    get "http://example.com/"
    assert_equal last_response.status, 200
  end

  def test_predicate_true
    self.app = Rack::WWW.new(default_app, :predicate => lambda { |env| true } )
    get "http://example.com/"
    assert_equal "http://www.example.com/", last_response.headers['Location']
  end
end


lib/rack/www.rb
require 'rack'
require 'rack/request'
require 'ipaddr'

module Rack
  # Rack::WWW
  class WWW
    def initialize(app, options = {})
      @options = { subdomain: 'www' }.merge(options)
      @app = app

      @redirect = !@options[:www].nil? ? @options[:www] : true
      @message = @options[:message]
      @subdomain = @options[:subdomain]
      @host_regex = @options[:host_regex] || /.+/i
      @predicate = @options[:predicate]
    end

    def call(env)
      if redirect?(env)
        redirect(env)
      else
        @app.call(env)
      end
    end

    private

    def redirect(env)
      url = prepare_url(env)
      headers = { 'Content-Type' => 'text/html', 'location' => url }

      if @message.respond_to?(:each)
        message = @message
      else
        message = [@message || '']
      end
      [301, headers, message]
    end

    def redirect?(env)
      predicate?(env) && change_subdomain?(env) && !ip_request?(env) &&
        matches_host?(env)
    end

    def predicate?(env)
      if @predicate
        @predicate.call env
      else
        true
      end
    end

    def change_subdomain?(env)
      @redirect && !already_subdomain?(env) ||
        !@redirect && already_subdomain?(env)
    end

    def ip_request?(env)
      IPAddr.new(env['SERVER_NAME'].to_s
        .gsub(/^(#{@subdomain}\.)/, '')
        .gsub(/^(www\.)/, ''))
    rescue IPAddr::InvalidAddressError
      false
    end

    def already_subdomain?(env)
      env['HTTP_HOST'].to_s.downcase =~ /^(#{@subdomain}\.)/
    end

    def matches_host?(env)
      env['HTTP_HOST'].to_s.downcase =~ @host_regex
    end

    def prepare_url(env)
      scheme = env['rack.url_scheme']
      host, port, path, query_string = extract_host(env)

      if @redirect == true
        host = "://#{@subdomain}." + host
      else
        host = '://' + host
      end

      "#{scheme}#{host}#{port}#{path}#{query_string}"
    end

    def extract_host(env)
      [env['SERVER_NAME'].to_s.gsub(/^(#{@subdomain}\.|www\.)/, ''),
       %w(80 443).include?(env['SERVER_PORT']) ? '' : ":#{env['SERVER_PORT']}",
       env['PATH_INFO'],
       env['QUERY_STRING'].empty? ? '' : '?' + env['QUERY_STRING'].to_s]
    end
  end
end
